File: extractors/typescript/src/model/writer.ts
import { compareKeys, compareText, isModuleKey, keyIndex, renderKey, type Key } from "./keys.js";
import type { Anchor, Edge, Entity, Literal, Model, NamedArgument } from "./model.js";

/**
 * The JSONL writer (schemas/README.md): one record per line in canonical
 * order, every reference a surrogate, every path interned, every closed
 * vocabulary an index into the header's dictionaries. Written to match
 * core's own encoder byte for byte — JSON.stringify IS the reference
 * serializer, so the only thing to get right is the order of everything.
 */

const SCHEMA_VERSION = "1.0.0";

/** The key order every entity record is written in (core's ENTITY_KEY_ORDER). */
const ENTITY_KEY_ORDER = [
  "name",
  "signature",
  "declaredType",
  "isStub",
  "parent",
  "attachedTo",
  "parameters",
  "localVariables",
  "definedIn",
  "comments",
  "metrics",
  "value",
  "space",
  "anchor",
] as const;

export class ModelWriteError extends Error {
  override name = "ModelWriteError";
}

function sortedDistinct(values: Iterable<string>): string[] {
  return [...new Set(values)].sort(compareText);
}

/** Model → lines, canonical order. The caller adds the LF and decides where bytes go. */
export function* encode(model: Model): Generator<string> {
  // 1. Canonical order is the surrogate assignment.
  const entities = [...model.entities].sort((a, b) => compareKeys(a.key, b.key));
  const surrogateOf = new Map<string, number>();
  entities.forEach((entity, index) => {
    const at = keyIndex(entity.key);
    if (surrogateOf.has(at)) throw new ModelWriteError(`duplicate natural key: ${renderKey(entity.key)}`);
    surrogateOf.set(at, index);
  });

  // 2. A module names itself: the entity with an empty, undisambiguated symbol.
  const moduleSurrogate = new Map<string, number>();
  entities.forEach((entity, index) => {
    if (isModuleKey(entity.key) && !moduleSurrogate.has(entity.key.module)) {
      moduleSurrogate.set(entity.key.module, index);
    }
  });

  const refOf = (key: Key, from: string): number => {
    const surrogate = surrogateOf.get(keyIndex(key));
    if (surrogate === undefined) {
      throw new ModelWriteError(`${from} references an entity this model does not declare: ${renderKey(key)}`);
    }
    return surrogate;
  };

  // 3. The file table: every path any record can name, interned once, sorted.
  const paths = new Set<string>();
  for (const entity of entities) {
    if (entity.anchor !== undefined) paths.add(entity.anchor.file);
    for (const path of entity.definedIn ?? []) paths.add(path);
  }
  for (const edge of model.edges) {
    paths.add(edge.anchor.file);
    if (edge.sourceFile !== undefined) paths.add(edge.sourceFile);
  }
  const files = [...paths].sort(compareText);
  const fileIndex = new Map(files.map((path, index) => [path, index]));
  const fileRef = (path: string): number => {
    const index = fileIndex.get(path);
    if (index === undefined) throw new ModelWriteError(`unknown path: ${path}`);
    return index;
  };
  const wireAnchor = (anchor: Anchor): [number, number, number] => [
    fileRef(anchor.file),
    anchor.span[0],
    anchor.span[1],
  ];

  // 4. Dictionaries: what this model actually uses, sorted.
  const dict = {
    kinds: sortedDistinct(entities.map((entity) => entity.kind)),
    traits: sortedDistinct(entities.flatMap((entity) => entity.traits)),
    edges: sortedDistinct(model.edges.map((edge) => edge.kind)),
    provenance: sortedDistinct(model.edges.map((edge) => edge.provenance)),
  };
  const indexOf = (values: readonly string[]): ((value: string) => number) => {
    const map = new Map(values.map((value, index) => [value, index]));
    return (value) => map.get(value) as number;
  };
  const kindRef = indexOf(dict.kinds);
  const traitRef = indexOf(dict.traits);
  const edgeKindRef = indexOf(dict.edges);
  const provenanceRef = indexOf(dict.provenance);

  yield JSON.stringify({
    t: "header",
    schemaVersion: SCHEMA_VERSION,
    lang: model.lang,
    extractor: model.extractor,
    root: model.root,
    ...(model.repository === undefined ? {} : { repository: model.repository }),
    dict,
  });

  for (const [index, path] of files.entries()) {
    yield JSON.stringify({ t: "f", i: index, path });
  }

  for (const [index, entity] of entities.entries()) {
    const moduleIndex = moduleSurrogate.get(entity.key.module);
    if (moduleIndex === undefined) {
      throw new ModelWriteError(
        `entity ${renderKey(entity.key)} names module "${entity.key.module}", which declares no module entity`,
      );
    }
    const record: Record<string, unknown> = {
      t: "e",
      i: index,
      k: kindRef(entity.kind),
      tr: entity.traits.map(traitRef),
      m: moduleIndex,
      // A module writes its own path in the symbol slot; everything else its
      // path below the module — empty for a nameless top-level entity.
      s: index === moduleIndex ? entity.key.module : entity.key.symbol,
    };
    if (entity.key.disambiguator !== undefined) record["d"] = entity.key.disambiguator;

    const owner = renderKey(entity.key);
    for (const wireKey of ENTITY_KEY_ORDER) {
      const value = entity[wireKey];
      if (value === undefined) continue;
      record[wireKey] = encodeValue(wireKey, value, owner);
    }
    yield JSON.stringify(record);
  }

  const edges = model.edges
    .map((edge) => ({
      edge,
      f: refOf(edge.from, `edge ${edge.kind}`),
      o: refOf(edge.to, `edge ${edge.kind}`),
    }))
    .sort(
      (a, b) =>
        a.f - b.f ||
        a.o - b.o ||
        compareText(a.edge.kind, b.edge.kind) ||
        fileRef(a.edge.anchor.file) - fileRef(b.edge.anchor.file) ||
        a.edge.anchor.span[0] - b.edge.anchor.span[0] ||
        a.edge.anchor.span[1] - b.edge.anchor.span[1] ||
        compareText(a.edge.provenance, b.edge.provenance),
    );

  for (const { edge, f, o } of edges) {
    const record: Record<string, unknown> = {
      t: "x",
      k: edgeKindRef(edge.kind),
      f,
      o,
      p: provenanceRef(edge.provenance),
    };
    if (edge.candidates !== undefined) {
      record["candidates"] = edge.candidates.map((key) => refOf(key, "edge candidates"));
    }
    // Omitted when empty: the edge KIND already says the key is there.
    if (edge.arguments !== undefined && edge.arguments.length > 0) {
      record["arguments"] = encodeArguments(edge.arguments, "annotation argument");
    }
    if (edge.isRead !== undefined) record["isRead"] = edge.isRead;
    if (edge.isWrite !== undefined) record["isWrite"] = edge.isWrite;
    if (edge.sourceFile !== undefined) record["sourceFile"] = fileRef(edge.sourceFile);
    record["anchor"] = wireAnchor(edge.anchor);
    yield JSON.stringify(record);
  }

  yield JSON.stringify({
    t: "eof",
    counts: { files: files.length, entities: entities.length, edges: edges.length },
  });

  function encodeValue(key: string, value: unknown, owner: string): unknown {
    switch (key) {
      case "anchor":
        return wireAnchor(value as Anchor);
      case "definedIn":
        return (value as string[]).map(fileRef);
      case "metrics": {
        // Key-sorted: canonical order reaches inside the record.
        const metrics = value as Record<string, number>;
        const sorted: Record<string, number> = {};
        for (const name of Object.keys(metrics).sort(compareText)) sorted[name] = metrics[name] as number;
        return sorted;
      }
      case "value":
        return encodeLiteral(value as Literal, `${owner}.value`);
      case "declaredType":
      case "parent":
      case "attachedTo":
        return refOf(value as Key, `${owner}.${key}`);
      case "parameters":
      case "localVariables":
        return (value as Key[]).map((ref) => refOf(ref, `${owner}.${key}`));
      default:
        return value;
    }
  }

  function encodeLiteral(value: Literal, from: string): unknown {
    switch (value.k) {
      case "enum":
        return { k: "enum", type: refOf(value.type, from), name: value.name };
      case "type":
        return { k: "type", type: refOf(value.type, from) };
      case "array":
        return { k: "array", items: value.items.map((item) => encodeLiteral(item, from)) };
      case "annotation":
        return {
          k: "annotation",
          type: refOf(value.type, from),
          arguments: encodeArguments(value.arguments, from),
        };
      default:
        return value;
    }
  }

  function encodeArguments(args: readonly NamedArgument[], from: string): unknown[] {
    return args.map((argument) => ({ name: argument.name, value: encodeLiteral(argument.value, from) }));
  }
}

/** The whole file as one string — LF-terminated lines, UTF-8 when written. */
export function encodeToString(model: Model): string {
  let out = "";
  for (const line of encode(model)) out += `${line}\n`;
  return out;
}

/** How many lines `encode` will yield: header + files + entities + edges + eof. */
export function plannedRecords(model: Model): number {
  const paths = new Set<string>();
  for (const entity of model.entities) {
    if (entity.anchor !== undefined) paths.add(entity.anchor.file);
    for (const path of entity.definedIn ?? []) paths.add(path);
  }
  for (const edge of model.edges) {
    paths.add(edge.anchor.file);
    if (edge.sourceFile !== undefined) paths.add(edge.sourceFile);
  }
  return 2 + paths.size + model.entities.length + model.edges.length;
}

File: extractors/elixir/lib/codegraph_elixir/model/writer.ex
defmodule CodegraphElixir.Model.Writer do
  @moduledoc """
  The JSONL writer (schemas/README.md): one record per line in canonical
  order, every reference a surrogate, every path interned, every closed
  vocabulary an index into the header's dictionaries. Written to match
  core's own encoder byte for byte — the second independent encoder, so the
  only thing to get right is the order of everything.
  """

  alias CodegraphElixir.Model
  alias CodegraphElixir.Model.{Edge, Entity, Json, Key}

  @schema_version "1.0.0"

  defmodule WriteError do
    defexception [:message]
  end

  @doc "Model → lines (iodata, no LF), canonical order. The caller adds the LF and decides where bytes go."
  def encode(%Model{} = model) do
    # 1. Canonical order is the surrogate assignment.
    entities = Enum.sort_by(model.entities, &Key.sort_key(&1.key))

    surrogate_of =
      entities
      |> Enum.with_index()
      |> Enum.reduce(%{}, fn {entity, index}, acc ->
        at = Key.index(entity.key)

        if Map.has_key?(acc, at) do
          raise WriteError, "duplicate natural key: #{Key.render(entity.key)}"
        end

        Map.put(acc, at, index)
      end)

    # 2. A module names itself: the entity with an empty, undisambiguated symbol.
    module_surrogate =
      entities
      |> Enum.with_index()
      |> Enum.reduce(%{}, fn {entity, index}, acc ->
        if Key.module?(entity.key), do: Map.put_new(acc, entity.key.module, index), else: acc
      end)

    ref_of = fn %Key{} = key, from ->
      case Map.fetch(surrogate_of, Key.index(key)) do
        {:ok, surrogate} ->
          surrogate

        :error ->
          raise WriteError, "#{from} references an entity this model does not declare: #{Key.render(key)}"
      end
    end

    # 3. The file table: every path any record can name, interned once, sorted.
    files =
      Enum.flat_map(entities, fn entity ->
        anchor = if entity.anchor, do: [elem(entity.anchor, 0)], else: []
        anchor ++ (entity.defined_in || [])
      end)
      |> Kernel.++(Enum.map(model.edges, &elem(&1.anchor, 0)))
      |> sorted_distinct()

    file_index = files |> Enum.with_index() |> Map.new()

    file_ref = fn path ->
      case Map.fetch(file_index, path) do
        {:ok, index} -> index
        :error -> raise WriteError, "unknown path: #{path}"
      end
    end

    wire_anchor = fn {file, start_line, end_line} -> [file_ref.(file), start_line, end_line] end

    # 4. Dictionaries: what this model actually uses, sorted.
    kinds = sorted_distinct(Enum.map(entities, & &1.kind))
    traits = sorted_distinct(Enum.flat_map(entities, & &1.traits))
    edge_kinds = sorted_distinct(Enum.map(model.edges, & &1.kind))
    provenances = sorted_distinct(Enum.map(model.edges, & &1.provenance))
    index_of = fn values -> values |> Enum.with_index() |> Map.new() end
    kind_ref = index_of.(kinds)
    trait_ref = index_of.(traits)
    edge_kind_ref = index_of.(edge_kinds)
    provenance_ref = index_of.(provenances)

    header =
      Json.object(
        [
          {"t", "header"},
          {"schemaVersion", @schema_version},
          {"lang", model.lang},
          {"extractor", {:object, model.extractor}},
          {"root", model.root}
        ] ++
          if(model.repository == nil, do: [], else: [{"repository", {:object, model.repository}}]) ++
          [
            {"dict",
             {:object,
              [{"kinds", kinds}, {"traits", traits}, {"edges", edge_kinds}, {"provenance", provenances}]}}
          ]
      )

    file_lines =
      files
      |> Enum.with_index()
      |> Enum.map(fn {path, index} -> Json.object([{"t", "f"}, {"i", index}, {"path", path}]) end)

    entity_lines =
      entities
      |> Enum.with_index()
      |> Enum.map(fn {%Entity{} = entity, index} ->
        module_index =
          case Map.fetch(module_surrogate, entity.key.module) do
            {:ok, at} ->
              at

            :error ->
              raise WriteError,
                    "entity #{Key.render(entity.key)} names module \"#{entity.key.module}\", which declares no module entity"
          end

        owner = Key.render(entity.key)

        head =
          [
            {"t", "e"},
            {"i", index},
            {"k", Map.fetch!(kind_ref, entity.kind)},
            {"tr", Enum.map(entity.traits, &Map.fetch!(trait_ref, &1))},
            {"m", module_index},
            # A module writes its own path in the symbol slot; everything else
            # its path below the module.
            {"s", if(index == module_index, do: entity.key.module, else: entity.key.symbol)}
          ] ++ if(entity.key.d == nil, do: [], else: [{"d", entity.key.d}])

        trait_keys =
          [
            {"name", entity.name},
            {"signature", entity.signature},
            {"isStub", entity.is_stub},
            {"parent", entity.parent && ref_of.(entity.parent, owner <> ".parent")},
            {"attachedTo", entity.attached_to && ref_of.(entity.attached_to, owner <> ".attachedTo")},
            {"parameters",
             entity.parameters && Enum.map(entity.parameters, &ref_of.(&1, owner <> ".parameters"))},
            {"definedIn", entity.defined_in && Enum.map(entity.defined_in, file_ref)},
            {"comments", entity.comments},
            {"metrics", entity.metrics && {:object, sort_pairs(entity.metrics)}},
            {"anchor", entity.anchor && wire_anchor.(entity.anchor)}
          ]
          |> Enum.reject(fn {_, value} -> value == nil end)

        # Keys no trait contributes ride through, sorted after the trait keys.
        extras = sort_pairs(entity.extra)
        Json.object(head ++ trait_keys ++ extras)
      end)

    edges =
      model.edges
      |> Enum.map(fn %Edge{} = edge ->
        {edge, ref_of.(edge.from, "edge #{edge.kind}"), ref_of.(edge.to, "edge #{edge.kind}")}
      end)
      |> Enum.sort_by(fn {edge, f, o} ->
        {file, start_line, end_line} = edge.anchor
        {f, o, Key.utf16(edge.kind), file_ref.(file), start_line, end_line, Key.utf16(edge.provenance)}
      end)

    edge_lines =
      Enum.map(edges, fn {edge, f, o} ->
        Json.object(
          [
            {"t", "x"},
            {"k", Map.fetch!(edge_kind_ref, edge.kind)},
            {"f", f},
            {"o", o},
            {"p", Map.fetch!(provenance_ref, edge.provenance)}
          ] ++
            if(edge.candidates == nil,
              do: [],
              else: [{"candidates", Enum.map(edge.candidates, &ref_of.(&1, "edge candidates"))}]
            ) ++
            [{"anchor", wire_anchor.(edge.anchor)}]
        )
      end)

    eof =
      Json.object([
        {"t", "eof"},
        {"counts",
         {:object, [{"files", length(files)}, {"entities", length(entities)}, {"edges", length(edges)}]}}
      ])

    [header] ++ file_lines ++ entity_lines ++ edge_lines ++ [eof]
  end

  @doc "The whole file as one binary — LF-terminated lines."
  def encode_to_string(model) do
    model |> encode() |> Enum.map(&[&1, "\n"]) |> IO.iodata_to_binary()
  end

  defp sorted_distinct(values) do
    values |> Enum.uniq() |> Enum.sort_by(&Key.utf16/1)
  end

  defp sort_pairs(pairs) do
    Enum.sort_by(pairs, fn {key, _} -> Key.utf16(to_string(key)) end)
  end
end
